src/datetime_rotation.jl

```diff
@@ -50,15 +61,25 @@ function DatetimeRotatingFileLogger(f::Union{Function,Nothing}, dir, filename_pa
     end
     # abspath in case user constructs the logger with a relative path and later cd's.
     drfl = DatetimeRotatingFileLogger(logger, abspath(dir),
-        DateFormat(filename_pattern), now(), always_flush)
+        DateFormat(filename_pattern), now(), always_flush, ReentrantLock(), nothing, callback)
     reopen!(drfl)
     return drfl
 end
 
 similar_logger(::SimpleLogger, io) = SimpleLogger(io, BelowMinLevel)
 similar_logger(l::FormatLogger, io) = FormatLogger(l.f, io, l.always_flush)
 function reopen!(drfl::DatetimeRotatingFileLogger)
-    io = open(calc_logpath(drfl.dir, drfl.filename_pattern), "a")
+    if drfl.current_file !== nothing
+        # flush and close the old IOStream
+        flush(drfl.logger.stream)
+        close(drfl.logger.stream)
+        if drfl.callback !== nothing
+            drfl.callback(drfl.current_file)
```

Comment:
I wonder if we should default the callback to a no-op like `identity` so we can avoid this check?

Reply:
Sure, if you prefer.